Fix GUS resampling regression & emulate the GUS' output more authentically

johnnovak opened the pull request:
# Description

Previously, the output of the Gravis UltraSound was rendered at the GUS' internal mixer rate which is dependent on the number of active voices, then this output was LERP-upsampled to the host mixer rate, so typically to 44.1 or 48 kHz.

Typical GUS mixer rate examples:

- **Star Control II:** 28 active voices, 22050 Hz mixer sample rate

- **DOOM I & II:** 28 active voices, 22050 Hz mixer sample rate (with the maximum 8 sfx channels configured)

- **Pinball Fantasies:** 16 active voices, 38587 Hz mixer sample rate

This upsampling basically had a weird smoothing/low-pass filtering effect on the output which was not too far off from the GUS Classic sound _by fluke_.

My intention was to only keep the LERP upsampling for the `sb_filter = modern` mode, so nothing else would use it. But I made a mistake: I left the default resampling method at the LERP upsampler, so some audio devices like the GUS kept using it instead of the proper Speex resampler. I fixed the wrong default in this commit, and this is what introduced the "regression" (subtle change of behaviour is a better name for it): 76f6bd0744861d

So this changed the resampling method from LERP upsampling to Speex for the GUS, and that removed the high-end-smoothing effect of the LERP upsampler (if you use linear interpolation as a "poor man's upsampling method", that acts as wonky and slightly crappy lowpass filter). This made things sound subtly different in the high frequencies (subjectively, the top end became a bit harsher).

This PR introduces a better approach that gets us much closer to the actual GUS Classic sound:

- On real hardware, the GUS always outputs a 44.1 kHz sample stream to the DAC; the channel count dependent mixer rate degradation is achieved via dropping samples in the GUS mixer.

- Our GUS emulation renders at the mixer rate, but with ZOH upsampling to 44.1 kHz we can approximate this sample dropping and fixed 44.1 kHz output rate behaviour almost perfectly. We then resample this with Speex to the host rate which is usually 48 kHz.

- But this adds a lot of high-end, so we need to faithfully emulate the analog low-pass filter stage of the GUS as well. On real hardware, some gentle low-pass filtering starts from about 7-9 kHz; at 10kHz the drop is about -3 dB, and at 20kHz around -9 dB. A 6 dB/oct LPF from 8 kHz gets us very close to this sound (this has been tweaked by comparing the filtered output to real hardware GUS Classic recordings).

---

I'm also enabling the GUS by default in this PR. Why? Because the GUS is cool! 😎 It's configured not to conflict with the SB, should this shouldn't cause any problems with existing configs. @kklobe rejoice 🥳 

CPU usage is a non-issue as all audio channels are put to sleep after a few seconds of inactivity. So you'll only "pay for it" when something actually uses the GUS.

@Python-Exoproject You might wanna play it extra safe and add `gus = false` to your global config to disable the GUS by default.

## Reference recordings

- [Pinball Fantasies (GUS Classic)](https://www.youtube.com/watch?v=SyMMxKJN_cg)
- [Doom II (GUS Classic)](https://www.youtube.com/watch?v=yMZi-QPWnwU)
- [Star Control 2 (GUS ACE)](https://nerdlypleasures.blogspot.com/2016/10/reasons-for-owning-gravis-ultrasound.html)

## Other references

- [Gravis UltraSound frequency response](https://www.epanorama.net/documents/soundcardtest/gus_classic.html#)

From the article:

![gusout](https://github.com/user-attachments/assets/d1592ddb-9ae4-4854-9b67-b48297fd9d73)


## Related issues

Fixes https://github.com/dosbox-staging/dosbox-staging/issues/4212


# Release notes

### Gravis UltraSound changes

- The Gravis UltraSound is now enabled by default.

- Fix subtle Gravis UltraSound regression where the high-frequency content of the GUS's output was a little harsher than before. This was mostly only noticeable in music with lots of hi-hats and similar high-pitched sounds.

- Additionally, the sound of the Gravis UltraSound Classic and MAX models are emulated more faithfully now. This includes the emulation of the analog low-pass output filter which is enabled by default.


# Manual testing

Tested the following games and compared the output to real hardware recordings:

- Doom I
- Doom II
- Pinball Fantasies
- Star Control 2


The change has been manually tested on:

- [x] Windows
- [ ] macOS
- [ ] Linux


# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] my change has been manually tested on Windows, macOS, and Linux.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [x] provided the release notes draft (for significant user-facing changes).
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

weirddan455 commented:
Nice job figuring this out. Do you have a specific commit for when this regression happened? I like to document that when possible and it might help me understand more. These upsampling and filtering options go a bit over my head (there's a lot of different combinations we use).

>  The Gravis UltraSound is now enabled by default.

My main concern with that is that for the majority of games, GUS doesn't work without placing drivers in `C:\ULTRASND`. I'm worried some games will auto-detect GUS and then leave users wondering why they have no sound in the new version where it previously worked with SoundBlaster or something.

IMO enabled by default features should be for things that work without additional steps. If the user has to manually install drivers, they should also be able to change a config setting.

johnnovak commented:
> Nice job figuring this out. Do you have a specific commit for when this regression happened? I like to document that when possible and it might help me understand more. These upsampling and filtering options go a bit over my head (there's a lot of different combinations we use).

Thanks, I've updated the top comment with that info, plus provided a better explanation of the sequence of events. I've also updated the commit message with the same info.

> My main concern with that is that for the majority of games, GUS doesn't work without placing drivers in C:\ULTRASND. I'm worried some games will auto-detect GUS and then leave users wondering why they have no sound in the new version where it previously worked with SoundBlaster or something.

I'll reflect to this later today. Good that you're thinking about these things, so it's great that you brought it up. Just I have some further-reaching ideas on solving this and the question of ROMs in general.

weirddan455 commented on src/hardware/gus.cpp:
```diff
@@ -1571,10 +1606,11 @@ void init_gus_dosbox_settings(Section_prop& secprop)
 	assert(str_prop);
 	str_prop->Set_help(
 	        "Filter for the Gravis UltraSound audio output:\n"
+	        "  on:        Filter the output (default).\n"
 	        "  off:       Don't filter the output (default).\n"
```
Remove the `(default)` from "off"

johnnovak replied:
Cheers, fixed.

weirddan455 reviewed (approved):
Thanks for the clarification. The code looks fine. One minor comment about the config text.

It's your call ultimately what you want to do with the GUS default. Just saying we may get some complaints if people's sound stops working due to missing drivers.

johnnovak commented:
> Thanks for the clarification. The code looks fine. One minor comment about the config text.

Thanks man, fixed.
 
> It's your call ultimately what you want to do with the GUS default. Just saying we may get some complaints if people's sound stops working due to missing drivers.

Reverted that for now. I'm intending to experiment with including a minimal GUS folder in our Z drive (so with most EXE files stripped out, just the bare minimium for the 99% use cases).
